core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java

```diff
@@ -0,0 +1,238 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.rpc;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+/**
+ * Factory for creating instances of the AccumuloProtocol.
+ * <p>
+ * This protocol includes a custom header to ensure compatibility between different versions of the
+ * protocol.
+ */
+public class AccumuloProtocolFactory extends TCompactProtocol.Factory {
+
+  private static final long serialVersionUID = 1L;
+
+  private final boolean isClient;
+
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    static final byte PROTOCOL_VERSION = 1;
+    private static final boolean HEADER_HAS_TRACE = true;
+
+    private final boolean isClient;
+
+    private Span span = null;
+    private Scope scope = null;
+    private final Map<String,String> traceHeaders = new HashMap<>();
+
+    public AccumuloProtocol(TTransport transport, boolean isClient) {
+      super(transport);
+      this.isClient = isClient;
+    }
+
+    /**
+     * For client calls, add RPC span and write the validation header
+     */
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (!this.isClient) {
+        super.writeMessageBegin(message);
+      } else {
+        span = TraceUtil.startClientRpcSpan(this.getClass(), message.name);
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
+          super.writeMessageBegin(message);
+        } catch (TException e) {
+          if (span != null) {
+            TraceUtil.setException(span, e, false);
+          }
+          if (scope != null) {
+            scope.close();
+          }
+          if (span != null) {
+            span.end();
+          }
+          throw e;
+        }
+      }
+    }
+
+    /**
+     * Writes the Accumulo protocol header containing version and identification info
+     */
+    private void writeHeader() throws TException {
+      super.writeI32(MAGIC_NUMBER);
+      super.writeByte(PROTOCOL_VERSION);
+
+      if (this.isClient && span != null && span.getSpanContext().isValid()) {
```

Comment:
This method is only called when isClient is true, so the check seems misleading and unnecessary.